### ISSUE TITLE
Question about ADC reading and code structure.

### ISSUE THREAD

**Issue by AndyEveritt** [zapta/simple_stepper_motor_analyzer#10]
I can not currently source the ACS70331EOLCTR-2P5B3 current sensors so I have been looking for alternatives.

I found this sensor from TI https://www.ti.com/lit/ds/symlink/tmcs1101-q1.pdf?HQS=dis-dk-null-mousermode-dsf-pf-null-wwe&ts=1641839042153 which seems to be approximately the same, but the sensitivity for the 3.3V version only goes up to 200mV/A (instead of 400mV/A).

To make sure that I could change the source code to accept this sensor I looked for where the sensitivity and and current range is set. Potentially I am misunderstanding but the datasheet for the current sensor says its +/-2.5A with 400mV/A and 1.5V at 0A.
![image](https://user-images.githubusercontent.com/38423143/153370749-4ee8533b-8444-4d1b-8d4e-b1256cacb1cf.png)

In the code `src/misc/hardware_config.h`, the sensor struct appears to be defined
```cpp
struct SensorSpec {
  const char* name;
  const uint16_t range_milliamps;
  const float volts_per_amp;
  // These computed values are cached to speed up unit conversions.
  const float counts_per_amp;
  const float amps_per_count;
  const float milliamps_per_count;

  SensorSpec(const char* name_, uint16_t range_milliamps_, float volts_per_amp_)
      : name(name_),
        range_milliamps(range_milliamps_),
        volts_per_amp(volts_per_amp_),
        counts_per_amp(volts_per_amp * 4096 / 3.3),
        amps_per_count(1 / counts_per_amp),
        milliamps_per_count(1000 / counts_per_amp) {}

  // Convert adc value to milliamps.
  int adc_value_to_milliamps(int adc_value) const {
    return (int)(adc_value * milliamps_per_count);
  }

  // Convert adc value to amps.
  float adc_value_to_amps(int adc_value) const {
    return ((float)adc_value) * amps_per_count;
  }
};
```

As far as I can tell the conversion of adc value to (milli)amps does not take into account the 1.5V @ 0A offset that the sensor has?

For example, you create an instance of this struct using `static const SensorSpec GMR_2P5_SENSOR("G2P5A", 2500, 0.4);`. If the sensor was measuring 0A, it would output 1.5V, I believe this would read 2048 on the ADC (if you are using the an external 3.0V shunt reference for the ADC_VREF pin to improve ADC performance, which I can't see in the schematic either so might be a nice addition? Otherwise it will read ~1862).  

Having just gone even deeper into trying to understand the code, I have now found that there is an offset for each of the ADC's in `src/aquisition/analyzer.h` which is applied to the value before it is converted to (milli)amps, which answers my question about if the sensor offset was being accounted for.

My new question (based on trying to get more familiar with embedded programming by learning from others) is why have you split the offset and other current sensor specific variables across 2 structs? I assume it is because the offset varies sensor to sensor so you need to calibrate it dynamically which would mean you couldn't use `const` for the sensor definition, which I assume is for some minor speed gain?

I appoligse for the long and convoluted question. I started typing as I was discovering to keep track of my train of thought. In essence, I would just like to understand why you have structured the code the way you have so I can learn and either apply or avoid it in my own coding.

Also it may be interesting to apply the recommended change to the circuit to improve the ADC performance
![image](https://user-images.githubusercontent.com/38423143/153390149-29c99252-8ed9-4994-a8cb-0427f0942dbe.png)

**Comment by globalcitizen** [comment]
The TI part is __hall effect__ based. The originally specified part was __GMR__. There are also __TMR__ parts. [This paper](https://crocus-technology.com/wp-content/uploads/2019/03/AN117-From-Hall-Effect-to-TMR-Rev0.1.pdf) and [this site](https://www.electronicdesign.com/industrial-automation/article/21805210/whats-the-difference-between-tmr-and-gmr-sensors) explain the differences. Based on that, maybe it would be best to look for TMR parts and [these guys](https://www.sinomags.com/En.html) seem promising.

**Comment by zapta** [comment]
In another forum, somebody suggested this one
https://www.mouser.it/ProductDetail/ACEINNA/MCA1101-5-3?qs=PzGy0jfpSMs%252BNw%2FnqBFIdA%3D%3D
.  May require configuration tweaks because it has a different A/V
sensitivity.

It's difficult to find parts these days.

**Comment by globalcitizen** [comment]
Yep. __MCA1101-5-3__ is AMR based, according to [the datasheet](https://www.mouser.it/datasheet/2/940/6020-1104-01_B_Datasheet_2c_Current_Sensor_Fix_Gai-1622847.pdf). This 2015 review ([part one](http://magnetism.eu/esm/2015/slides/tiusan-slides1.pdf), [part two](http://magnetism.eu/esm/2015/slides/tiusan-slides2.pdf) compares AMR to the other approaches in more quantum physical detail than my mathematical literacy will stand for. But the result seems to be: if it's not TMR, it's functionally prior-generation and therefore potentially not the best thing to design for in terms of design longevity.
